File: src/main/java/org/taktik/icure/logic/impl/ClassificationTemplateLogicImpl.java
/*
 * Copyright (C) 2018 Taktik SA
 *
 * This file is part of iCureBackend.
 *
 * iCureBackend is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as published by
 * the Free Software Foundation.
 *
 * iCureBackend is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with iCureBackend.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.taktik.icure.logic.impl;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.taktik.icure.dao.ClassificationTemplateDAO;
import org.taktik.icure.dao.impl.idgenerators.UUIDGenerator;
import org.taktik.icure.db.PaginatedList;
import org.taktik.icure.db.PaginationOffset;
import org.taktik.icure.entities.ClassificationTemplate;
import org.taktik.icure.entities.embed.Delegation;
import org.taktik.icure.logic.ClassificationTemplateLogic;
import org.taktik.icure.logic.ICureSessionLogic;
import org.taktik.icure.validation.aspect.Check;

import javax.validation.constraints.NotNull;
import java.util.*;


/**
 * Created by dlm on 16-07-18
 */
@org.springframework.stereotype.Service
public class ClassificationTemplateLogicImpl extends GenericLogicImpl<ClassificationTemplate, ClassificationTemplateDAO> implements ClassificationTemplateLogic {
	private static final Logger log = LoggerFactory.getLogger(ClassificationTemplateLogicImpl.class);


	private ClassificationTemplateDAO classificationTemplateDAO;
	private UUIDGenerator uuidGenerator;
	private ICureSessionLogic sessionLogic;

	@Autowired
	public void setClassificationTemplateDAO(ClassificationTemplateDAO classificationTemplateDAO) {
		this.classificationTemplateDAO = classificationTemplateDAO;
	}
	@Autowired
	public void setUuidGenerator(UUIDGenerator uuidGenerator) {
		this.uuidGenerator = uuidGenerator;
	}
	@Autowired
	public void setSessionLogic(ICureSessionLogic sessionLogic) {
		this.sessionLogic = sessionLogic;
	}
	@Override
	protected ClassificationTemplateDAO getGenericDAO() {
		return classificationTemplateDAO;
	}

	@Override
	public ClassificationTemplate createClassificationTemplate(@Check @NotNull ClassificationTemplate classificationTemplate) {
		List<ClassificationTemplate> createdClassificationTemplates = new ArrayList<>(1);
		try {
			// Fetching the hcParty
			String healthcarePartyId = sessionLogic.getCurrentHealthcarePartyId();

			// Setting Classification Template attributes
			classificationTemplate.setId(uuidGenerator.newGUID().toString());
			classificationTemplate.setAuthor(healthcarePartyId);
			classificationTemplate.setResponsible(healthcarePartyId);

			createEntities(Collections.singleton(classificationTemplate), createdClassificationTemplates);
		} catch (Exception e) {
			log.error("createClassificationTemplate: " + e.getMessage());
			throw new IllegalArgumentException("Invalid Classification Template", e);
		}
		return createdClassificationTemplates.size() == 0 ? null:createdClassificationTemplates.get(0);
	}

	@Override
	public ClassificationTemplate getClassificationTemplate(String classificationTemplateId) {
		return classificationTemplateDAO.getClassificationTemplate(classificationTemplateId);
	}


	@Override
	public Set<String> deleteClassificationTemplates(Set<String> ids) {
		try {
			deleteEntities(ids);
		} catch (Exception e) {
			log.error(e.getMessage(), e);
			return null;
		}
		return ids;
	}

	@Override
	public ClassificationTemplate modifyClassificationTemplate(@Check @NotNull ClassificationTemplate classificationTemplate) {
		try {
			ClassificationTemplate toEdit = getClassificationTemplate(classificationTemplate.getId());
			toEdit.setLabel(classificationTemplate.getLabel());
			updateEntities(Collections.singleton(toEdit));
			return getClassificationTemplate(classificationTemplate.getId());
		} catch (Exception e) {
			throw new IllegalArgumentException("Invalid Classification Template", e);
		}
	}

	@Override
	public ClassificationTemplate addDelegation(String classificationTemplateId, String healthcarePartyId, Delegation delegation) {
		ClassificationTemplate classificationTemplate = getClassificationTemplate(classificationTemplateId);
		classificationTemplate.addDelegation(healthcarePartyId, delegation);

		return classificationTemplateDAO.save(classificationTemplate);
	}

	@Override
	public ClassificationTemplate addDelegations(String classificationTemplateId, List<Delegation> delegations) {
		ClassificationTemplate classificationTemplate = getClassificationTemplate(classificationTemplateId);
		delegations.forEach(d->classificationTemplate.addDelegation(d.getDelegatedTo(),d));

		return classificationTemplateDAO.save(classificationTemplate);

	}

    @Override
    public List<ClassificationTemplate> getClassificationTemplateByIds(List<String> ids) {
        return classificationTemplateDAO.getList(ids);
    }

    @Override
    public List<ClassificationTemplate> findByHCPartySecretPatientKeys(String hcPartyId, ArrayList<String> secretPatientKeys) {
        return classificationTemplateDAO.findByHCPartySecretPatientKeys(hcPartyId, secretPatientKeys);
    }

	@Override
	public PaginatedList<ClassificationTemplate> listClassificationTemplates(PaginationOffset paginationOffset) {
		return classificationTemplateDAO.listClassificationTemplates(paginationOffset);
	}

}

File: src/main/kotlin/org/taktik/icure/asynclogic/impl/ClassificationTemplateLogicImpl.kt
/*
 *  iCure Data Stack. Copyright (c) 2020 Taktik SA
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU Affero General Public License as
 *     published by the Free Software Foundation, either version 3 of the
 *     License, or (at your option) any later version.
 *
 *     This program is distributed in the hope that it will be useful, but
 *     WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *     Affero General Public License for more details.
 *
 *     You should have received a copy of the GNU Affero General Public
 *     License along with this program.  If not, see
 *     <https://www.gnu.org/licenses/>.
 */
package org.taktik.icure.asynclogic.impl

import kotlinx.coroutines.ExperimentalCoroutinesApi
import kotlinx.coroutines.flow.*
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Service
import org.taktik.couchdb.DocIdentifier
import org.taktik.couchdb.ViewQueryResultEvent
import org.taktik.couchdb.id.UUIDGenerator
import org.taktik.icure.asyncdao.ClassificationTemplateDAO
import org.taktik.icure.asynclogic.AsyncSessionLogic
import org.taktik.icure.asynclogic.ClassificationTemplateLogic
import org.taktik.icure.db.PaginationOffset
import org.taktik.icure.entities.ClassificationTemplate
import org.taktik.icure.entities.embed.Delegation

/**
 * Created by dlm on 16-07-18
 */
@ExperimentalCoroutinesApi
@Service
class ClassificationTemplateLogicImpl(private val classificationTemplateDAO: ClassificationTemplateDAO,
                                      private val uuidGenerator: UUIDGenerator,
                                      private val sessionLogic: AsyncSessionLogic) : GenericLogicImpl<ClassificationTemplate, ClassificationTemplateDAO>(sessionLogic), ClassificationTemplateLogic {

    override fun getGenericDAO(): ClassificationTemplateDAO {
        return classificationTemplateDAO
    }

    override suspend fun createClassificationTemplate(classificationTemplate: ClassificationTemplate) = fix(classificationTemplate) { classificationTemplate ->
        try { // Fetching the hcParty
            val userId = sessionLogic.getCurrentUserId()
            val healthcarePartyId = sessionLogic.getCurrentHealthcarePartyId()
            // Setting Classification Template attributes
            createEntities(setOf(classificationTemplate.copy(
                    author = userId, responsible = healthcarePartyId
            ))).firstOrNull()
        } catch (e: Exception) {
            log.error("createClassificationTemplate: " + e.message)
            throw IllegalArgumentException("Invalid Classification Template", e)
        }
    }

    override suspend fun getClassificationTemplate(classificationTemplateId: String): ClassificationTemplate? {
        return classificationTemplateDAO.getClassificationTemplate(classificationTemplateId)
    }

    override fun deleteClassificationTemplates(ids: Set<String>): Flow<DocIdentifier> {
        return try {
            deleteEntities(ids)
        } catch (e: Exception) {
            log.error(e.message, e)
            flowOf()
        }
    }

    override suspend fun modifyClassificationTemplate(classificationTemplate: ClassificationTemplate) = fix(classificationTemplate) { classificationTemplate ->
        try {
            getClassificationTemplate(classificationTemplate.id)?.let { toEdit ->
                modifyEntities(setOf(toEdit.copy(label = classificationTemplate.label))).firstOrNull()
            } ?: throw IllegalArgumentException("Non-existing Classification Template")
        } catch (e: Exception) {
            throw IllegalArgumentException("Invalid Classification Template", e)
        }
    }

    override suspend fun addDelegation(classificationTemplateId: String, healthcarePartyId: String, delegation: Delegation): ClassificationTemplate? {
        val classificationTemplate = getClassificationTemplate(classificationTemplateId)
        return classificationTemplate?.let {
            classificationTemplateDAO.save(it.copy(delegations = it.delegations + mapOf(
                    healthcarePartyId to setOf(delegation)
            )))
        }
    }

    override suspend fun addDelegations(classificationTemplateId: String, delegations: List<Delegation>): ClassificationTemplate? {
        val classificationTemplate = getClassificationTemplate(classificationTemplateId)
        return classificationTemplate?.let {
            return classificationTemplateDAO.save(it.copy(
                    delegations = it.delegations +
                            delegations.mapNotNull { d -> d.delegatedTo?.let { delegateTo -> delegateTo to setOf(d) } }
            ))
        }
    }

    override fun getClassificationTemplates(ids: List<String>): Flow<ClassificationTemplate> = flow {
        emitAll(classificationTemplateDAO.getEntities(ids))
    }

    override fun listClasificationsByHCPartyAndSecretPatientKeys(hcPartyId: String, secretPatientKeys: ArrayList<String>): Flow<ClassificationTemplate> = flow {
        emitAll(classificationTemplateDAO.listClassificationsByHCPartyAndSecretPatientKeys(hcPartyId, secretPatientKeys))
    }

    override fun listClassificationTemplates(paginationOffset: PaginationOffset<String>) =flow<ViewQueryResultEvent> {
        emitAll(classificationTemplateDAO.findClassificationTemplates(paginationOffset))
    }

    companion object {
        private val log = LoggerFactory.getLogger(ClassificationTemplateLogicImpl::class.java)
    }
}
